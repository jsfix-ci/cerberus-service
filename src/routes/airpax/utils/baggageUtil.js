import * as pluralise from 'pluralise';
import { UNKNOWN_TEXT } from '../../../constants';

const getBaggageWeight = (baggage) => {
  if (!baggage?.weight) {
    return UNKNOWN_TEXT;
  }
  return baggage.weight;
};

const hasBaggage = (targetTask) => {
  return !!targetTask.movement.baggage;
};

const getBaggage = (targetTask) => {
  if (hasBaggage(targetTask)) {
    return targetTask.movement.baggage;
  }
  return null;
};

const getCheckedBags = (baggage, taskDetails = false) => {
  if ((!baggage || !baggage?.numberOfCheckedBags) && baggage?.numberOfCheckedBags !== 0) {
    return UNKNOWN_TEXT;
  }
  return pluralise.withCount(
    baggage.numberOfCheckedBags,
    '% checked bag',
    '% checked bags',
    (taskDetails ? 'None' : 'No checked bags'),
  );
};

const getNumberOfCheckedBags = (baggage) => {
  if ((!baggage || !baggage?.numberOfCheckedBags) && baggage?.numberOfCheckedBags !== 0) {
    return UNKNOWN_TEXT;
  }
  return baggage.numberOfCheckedBags;
};

const toFormattedCheckedBagsCount = (baggage) => {
  if ((!baggage || !baggage?.numberOfCheckedBags) && baggage?.numberOfCheckedBags !== 0) {
    return UNKNOWN_TEXT;
  }
  return pluralise.withCount(
    baggage.numberOfCheckedBags,
    '% bag',
    '% bags',
    'None',
  );
};

const BaggageUtil = {
  get: getBaggage,
  weight: getBaggageWeight,
  checked: getCheckedBags,
  checkedCount: getNumberOfCheckedBags,
  formatCheckedCount: toFormattedCheckedBagsCount,
};

export default BaggageUtil;

export { getBaggageWeight, getBaggage, getCheckedBags, getNumberOfCheckedBags };
