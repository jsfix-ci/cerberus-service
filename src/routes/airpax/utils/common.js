import lookup from 'country-code-lookup';
import { UNKNOWN_TEXT } from '../../../constants';

const convertToIso3Code = (iso2Code) => {
  if (!iso2Code) {
    return UNKNOWN_TEXT;
  }
  if (!lookup.byIso(iso2Code)) {
    return UNKNOWN_TEXT;
  }
  return lookup.byIso(iso2Code).iso3;
};

const Common = {
  iso3Code: convertToIso3Code,
};

export default Common;

export { convertToIso3Code };
