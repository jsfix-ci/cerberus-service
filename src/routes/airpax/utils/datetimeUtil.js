import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import relativeTime from 'dayjs/plugin/relativeTime';
import customParseFormat from 'dayjs/plugin/customParseFormat';
import { AGO_TEXT, UNKNOWN_TEXT, UTC_DATE_REGEXS } from '../../../constants';

dayjs.extend(utc);
dayjs.extend(relativeTime);
dayjs.extend(customParseFormat);

const validateDate = (date) => {
  return !!(date !== UNKNOWN_TEXT && date && UTC_DATE_REGEXS.some((regex) => regex.test(date)));
};

/**
 * This will return either a formatted UTC datetime or a dayjs datetime object.
 * @param {*} asDayjsObj Flag to determine whether to return a formatted UTC datetime or a dayjs datetime object.
 * @returns A UTC formatted time or a dayjs datetime object.
 */
const getDate = (asDayjsObj = false) => {
  if (asDayjsObj) {
    return dayjs.utc();
  }
  return dayjs.utc().format();
};

const toRelativeTime = (date) => {
  if (!validateDate(date)) {
    return UNKNOWN_TEXT;
  }
  const dateTimeStart = dayjs.utc(date);
  return dateTimeStart.fromNow();
};

const isInPast = (date) => {
  if (!validateDate(date)) {
    return UNKNOWN_TEXT;
  }
  return toRelativeTime(date)?.endsWith(AGO_TEXT);
};

const getFormattedDate = (date, dateFormat) => {
  if (!validateDate(date)) {
    return UNKNOWN_TEXT;
  }
  return dayjs.utc(date).format(dateFormat);
};

const formatToUTCDate = (date, inputFormat, outputFormat) => {
  if (!date) {
    return UNKNOWN_TEXT;
  }
  return dayjs.utc(date, inputFormat).format(outputFormat);
};

const toDateTimeList = (dateOne, dateTwo) => {
  return [dateOne, dateTwo];
};

const DateTimeUtil = {
  convertToUTC: formatToUTCDate,
  date: getDate,
  format: getFormattedDate,
  isPast: isInPast,
  toList: toDateTimeList,
  relativeTime: toRelativeTime,
  validate: validateDate,
};

export default DateTimeUtil;

export { formatToUTCDate, getDate, getFormattedDate, isInPast, toDateTimeList, toRelativeTime };
