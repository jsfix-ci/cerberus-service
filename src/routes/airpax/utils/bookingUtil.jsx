import React from 'react';
import lookup from 'country-code-lookup';
import { getFormattedDate, toDateTimeList } from './datetimeUtil';
import calculateTimeDifference from '../../../utils/calculateDatetimeDifference';

import {
  UNKNOWN_TEXT,
  SHORT_DATE_FORMAT_ALT,
  STANDARD_HOUR_MINUTE_FORMAT,
  STANDARD_CARD_EXPIRY_FORMAT,
} from '../../../constants';

const getAgentLocation = (agent) => {
  if (!agent?.location) {
    return UNKNOWN_TEXT;
  }
  return agent.location;
};

const getAgentIata = (agent) => {
  if (!agent?.iata) {
    return UNKNOWN_TEXT;
  }
  return agent.iata;
};

const getAgentName = (agent) => {
  if (!agent?.name) {
    return UNKNOWN_TEXT;
  }
  return agent.name;
};

const hasAgent = (booking) => {
  return !!booking?.agent;
};

const getAgent = (booking) => {
  if (hasAgent(booking)) {
    return booking.agent;
  }
  return null;
};

const hasPaymentCard = (payment) => {
  return !!payment?.card;
};

const getPaymentCard = (payment) => {
  if (hasPaymentCard(payment)) {
    return payment.card;
  }
  return null;
};

const getCardExpiry = (payment) => {
  const paymentCard = getPaymentCard(payment);
  if (paymentCard?.expiry) {
    return getFormattedDate(paymentCard.expiry, STANDARD_CARD_EXPIRY_FORMAT);
  }
  return UNKNOWN_TEXT;
};

const getCardLastFourDigits = (payment) => {
  const paymentCard = getPaymentCard(payment);
  if (paymentCard?.number) {
    return paymentCard.number.substr(paymentCard.number.length - 4);
  }
  return UNKNOWN_TEXT;
};

const getPaymentAmount = (payment) => {
  if (!payment?.amount) {
    return UNKNOWN_TEXT;
  }
  return payment.amount;
};

const hasPayments = (booking) => {
  return !!booking?.payments?.length > 0;
};

const getPayments = (booking) => {
  if (hasPayments(booking)) {
    return booking.payments;
  }
  return null;
};

const getTicketType = (ticket) => {
  if (!ticket?.type) {
    return UNKNOWN_TEXT;
  }
  return ticket.type;
};

const getTicketNumber = (ticket) => {
  if (!ticket?.number) {
    return UNKNOWN_TEXT;
  }
  return ticket.number;
};

const hasTicket = (booking) => {
  return !!booking?.ticket;
};

const getTicket = (booking) => {
  if (hasTicket(booking)) {
    return booking.ticket;
  }
  return null;
};

const getBookingType = (booking) => {
  if (!booking?.type) {
    return UNKNOWN_TEXT;
  }
  return booking.type;
};

const getBookingCountryCode = (booking) => {
  if (!booking?.country) {
    return UNKNOWN_TEXT;
  }
  return booking.country;
};

const getBookingCountryName = (booking) => {
  const countryCode = getBookingCountryCode(booking);
  if (countryCode === UNKNOWN_TEXT) {
    return UNKNOWN_TEXT;
  }
  if (lookup.byIso(countryCode) === null) {
    return UNKNOWN_TEXT;
  }
  return lookup.byIso(countryCode).country;
};

const toBookingDateText = (booking) => {
  if (!booking?.bookedAt) {
    return UNKNOWN_TEXT;
  }
  return getFormattedDate(booking.bookedAt, SHORT_DATE_FORMAT_ALT);
};

const getBookingReference = (booking) => {
  if (!booking?.reference) {
    return UNKNOWN_TEXT;
  }
  return booking.reference;
};

const getCheckInAt = (booking) => {
  if (!booking?.checkInAt) {
    return UNKNOWN_TEXT;
  }
  return booking.checkInAt;
};

const toCheckInTimeText = (booking) => {
  const checkinPrefix = 'Check-in';
  if (!booking?.checkInAt) {
    return `${checkinPrefix} ${UNKNOWN_TEXT}`;
  }
  return `${checkinPrefix} ${getFormattedDate(booking.checkInAt, STANDARD_HOUR_MINUTE_FORMAT)}`;
};

const getBookedAt = (booking) => {
  return booking?.bookedAt;
};

const getBookedPriorToDeparture = (bookedAt, departureTime) => {
  const dateTimeList = toDateTimeList(bookedAt, departureTime);
  return calculateTimeDifference(dateTimeList);
};

const hasBooking = (targetTask) => {
  return !!targetTask?.movement?.booking;
};

const getBooking = (targetTask) => {
  if (hasBooking(targetTask)) {
    return targetTask.movement.booking;
  }
  return null;
};

const toPaymentsBlock = (booking) => {
  const payments = getPayments(booking);
  if (payments) {
    return payments.map((payment, index) => {
      return (
        <div key={index} className="bottom-border-thin govuk-!-margin-top-1 govuk-!-padding-bottom-1">
          <div className="font__bold">{getPaymentAmount(payment)}</div>
          <div className="font__light">
            Credit card ending {getCardLastFourDigits(payment)}, expiry {getCardExpiry(payment)}
          </div>
        </div>
      );
    });
  }
};

const BookingUtil = {
  get: getBooking,
  bookedAt: getBookedAt,
  checkInAt: getCheckInAt,
  bookedPrior: getBookedPriorToDeparture,
  toCheckInText: toCheckInTimeText,
  bookingRef: getBookingReference,
  toBookingText: toBookingDateText,
  countryCode: getBookingCountryCode,
  countryName: getBookingCountryName,
  bookingType: getBookingType,
  bookingTicket: getTicket,
  ticketNumber: getTicketNumber,
  ticketType: getTicketType,
  payments: getPayments,
  containsPayments: hasPayments,
  paymentAmount: getPaymentAmount,
  paymentCard: getPaymentCard,
  cardLastFourDigits: getCardLastFourDigits,
  cardExpiry: getCardExpiry,
  agent: getAgent,
  agentName: getAgentName,
  agentIata: getAgentIata,
  agentLocation: getAgentLocation,
  paymentsBlock: toPaymentsBlock,
};

export default BookingUtil;

export {
  toBookingDateText,
  getBookingReference,
  toCheckInTimeText,
  getBooking,
  getBookedAt,
  getCheckInAt,
  getBookedPriorToDeparture,
  getBookingCountryCode,
  getBookingCountryName,
  getBookingType,
  getTicket,
  getTicketNumber,
  getTicketType,
  getPayments,
  getPaymentAmount,
  getPaymentCard,
  getCardLastFourDigits,
  getCardExpiry,
  getAgent,
  getAgentName,
  getAgentIata,
  getAgentLocation,
  toPaymentsBlock,
};
