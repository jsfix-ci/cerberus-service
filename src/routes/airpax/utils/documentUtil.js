import moment from 'moment';
import lookup from 'country-code-lookup';
import {
  A_TITLE_CASE_TEXT,
  AN_TITLE_CASE_TEXT,
  AGO_TEXT,
  A_SMALL_TEXT,
  AN_SMALL_TEXT,
  UNKNOWN_TEXT,
  AFTER_TRAVEL_TEXT,
  BEFORE_TRAVEL_TEXT,
  SHORT_DATE_FORMAT_ALT,
} from '../../../constants';
import { formatField } from '../../../utils/formatField';
import { getFormattedDate } from './datetimeUtil';

const calculateExpiry = (passportExpiry, arrivalTime) => {
  const expiry = arrivalTime
    && passportExpiry !== 'Unknown'
    && `${arrivalTime},${moment(passportExpiry).format('YYYY-MM-DDTHH:mm:ss')}`;
  if (expiry) {
    return formatField('BOOKING_DATETIME', expiry)
      .split(', ')[1]
      .replace(BEFORE_TRAVEL_TEXT, AFTER_TRAVEL_TEXT)
      .replace(AGO_TEXT, BEFORE_TRAVEL_TEXT)
      .replace(A_SMALL_TEXT, A_TITLE_CASE_TEXT)
      .replace(AN_SMALL_TEXT, AN_TITLE_CASE_TEXT);
  }
  return UNKNOWN_TEXT;
};

const hasDocument = (person) => {
  return !!person?.document;
};

const getDocument = (person) => {
  if (hasDocument(person)) {
    return person.document;
  }
  return null;
};

const getDocumentExpiry = (document, taskDetails = false) => {
  const expiryPrefix = 'Expires';
  if (!document?.expiry) {
    return taskDetails ? UNKNOWN_TEXT : `${expiryPrefix} ${UNKNOWN_TEXT}`;
  }
  return taskDetails ? `${getFormattedDate(document?.expiry, SHORT_DATE_FORMAT_ALT)}` : `${expiryPrefix} ${getFormattedDate(document?.expiry, SHORT_DATE_FORMAT_ALT)}`;
};

const getDocumentValidity = (document, taskDetails = false) => {
  const validityPrefix = 'Valid from';
  if (!document?.validFrom) {
    return taskDetails ? UNKNOWN_TEXT : `${validityPrefix} ${UNKNOWN_TEXT}`;
  }
  return taskDetails ? `${getFormattedDate(document?.validFrom, SHORT_DATE_FORMAT_ALT)}` : `${validityPrefix} ${getFormattedDate(document?.validFrom, SHORT_DATE_FORMAT_ALT)}`;
};

const getDocumentType = (document) => {
  if (!document?.type) {
    return UNKNOWN_TEXT;
  }
  return document.type;
};

const getDocumentNumber = (document) => {
  if (!document?.number) {
    return UNKNOWN_TEXT;
  }
  return document.number;
};

const getDocumentName = (person) => {
  if (!person?.name) {
    return UNKNOWN_TEXT;
  }
  const firstNames = person?.name?.first || UNKNOWN_TEXT;
  const lastName = person?.name?.last || UNKNOWN_TEXT;
  return `${lastName.toUpperCase()}, ${firstNames}`;
};

// TODO finish implementation once data flows through
const getDocumentIdentification = (document) => {
  if (!document) {
    return UNKNOWN_TEXT;
  }
  return UNKNOWN_TEXT;
};

const getDocumentCountryOfIssue = (document, taskDetails = false) => {
  const countryOfIssuePrefix = 'Issued by';
  if (!document?.countryOfIssue) {
    return taskDetails ? UNKNOWN_TEXT : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
  }
  if (taskDetails) {
    return lookup.byIso(document.countryOfIssue) !== null
      ? `${lookup.byIso(document.countryOfIssue).country} (${document.countryOfIssue})`
      : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
  }

  return lookup.byIso(document.countryOfIssue) !== null
    ? `${countryOfIssuePrefix} ${document.countryOfIssue}`
    : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
};

const getDocumentCountryOfIssueCode = (document) => {
  if (!document?.countryOfIssue) {
    return UNKNOWN_TEXT;
  }
  return document.countryOfIssue;
};

const getDocumentNationality = (document, taskDetails = false) => {
  return getDocumentCountryOfIssue(document, taskDetails);
};

const getDocumentDOB = (document) => {
  if (!document?.dateOfBirth) {
    return UNKNOWN_TEXT;
  }
  return getFormattedDate(document?.dateOfBirth, SHORT_DATE_FORMAT_ALT);
};

const DocumentUtil = {
  get: getDocument,
  docExpiry: getDocumentExpiry,
  docValidity: getDocumentValidity,
  docType: getDocumentType,
  docNumber: getDocumentNumber,
  docName: getDocumentName,
  docIdentification: getDocumentIdentification,
  docCountry: getDocumentCountryOfIssue,
  docCountryCode: getDocumentCountryOfIssueCode,
  docNationality: getDocumentNationality,
  docDOB: getDocumentDOB,
  calculateExpiry,
};

export default DocumentUtil;

export {
  getDocument,
  getDocumentExpiry,
  getDocumentValidity,
  getDocumentType,
  getDocumentNumber,
  getDocumentName,
  getDocumentIdentification,
  getDocumentCountryOfIssue,
  getDocumentCountryOfIssueCode,
  getDocumentNationality,
  getDocumentDOB,
  calculateExpiry,
};
