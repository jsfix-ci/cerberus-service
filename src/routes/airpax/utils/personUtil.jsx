import React from 'react';
import lookup from 'country-code-lookup';
import dayjs from 'dayjs';
import formatGender from '../../../utils/genderFormatter';
import { getFormattedDate } from './datetimeUtil';
import { SHORT_DATE_FORMAT_ALT, UNKNOWN_TEXT } from '../../../constants';

const getNationality = (person) => {
  if (!person?.nationality) {
    return UNKNOWN_TEXT;
  }
  return person.nationality;
};

const getCountryName = (person) => {
  if (!person?.nationality) {
    return UNKNOWN_TEXT;
  }
  return lookup.byIso(person.nationality).country;
};

const getDateOfBirth = (person) => {
  if (!person?.dateOfBirth) {
    return UNKNOWN_TEXT;
  }
  return getFormattedDate(person.dateOfBirth, SHORT_DATE_FORMAT_ALT);
};

const getAge = (person) => {
  if (!person?.dateOfBirth) {
    return UNKNOWN_TEXT;
  }
  return dayjs.utc().diff(dayjs(person.dateOfBirth), 'year');
};

const getGender = (person) => {
  if (!person) {
    return UNKNOWN_TEXT;
  }
  return formatGender(person.gender);
};

const getLastName = (person, capitalize = true) => {
  if (!person?.name?.last) {
    return capitalize ? UNKNOWN_TEXT.toUpperCase() : UNKNOWN_TEXT;
  }
  return capitalize ? person.name.last.toUpperCase() : person.name.last;
};

const getFirstName = (person) => {
  if (!person?.name?.first) {
    return UNKNOWN_TEXT;
  }
  return person.name.first;
};

const getFrequentFlyerNumber = (person) => {
  if (!person?.frequentFlyerNumber) {
    return UNKNOWN_TEXT;
  }
  return person.frequentFlyerNumber;
};

const hasSSRCodes = (person) => {
  return !!person?.ssrCodes?.length > 0;
};

const getSSRCodes = (person) => {
  if (!hasSSRCodes(person)) {
    return UNKNOWN_TEXT;
  }
  return person.ssrCodes.join(', ');
};

const toCoTravellers = (otherPersons) => {
  if (!otherPersons) {
    return <li className="govuk-!-font-weight-bold">None</li>;
  }
  const maxToDisplay = 4;
  const remaining = otherPersons.length > maxToDisplay ? otherPersons.length - maxToDisplay : 0;
  return otherPersons.map((person, index) => {
    if (index < maxToDisplay) {
      return (
        <li key={index} className="govuk-!-font-weight-bold">
          {getLastName(person)}, {getFirstName(person)}{(index !== maxToDisplay - 1)
                && (index !== otherPersons.length - 1) ? ',' : ''} {(remaining > 0 && index + 1 === maxToDisplay)
                  ? ` plus ${remaining} more` : ''}
        </li>
      );
    }
  });
};

const getAllPersons = (person, otherPersons) => {
  let allPersons = [];
  if (person) {
    allPersons.push(person);
  }
  if (otherPersons) {
    allPersons = allPersons.concat(otherPersons);
  }
  return allPersons;
};

const getTotalNumberOfPersons = (targetTask) => {
  if (!targetTask?.movement?.person) {
    return 0;
  }
  return targetTask.movement.otherPersons.length + 1;
};

const hasOtherPersons = (targetTask) => {
  return !!targetTask?.movement?.otherPersons?.length;
};

const getOtherPersons = (targetTask) => {
  if (hasOtherPersons(targetTask)) {
    return targetTask.movement.otherPersons;
  }
  return null;
};

const hasPerson = (targetTask) => {
  return !!targetTask?.movement?.person;
};

const getPerson = (targetTask) => {
  if (hasPerson(targetTask)) {
    return targetTask.movement.person;
  }
  return null;
};

const PersonUtil = {
  get: getPerson,
  getOthers: getOtherPersons,
  totalPersons: getTotalNumberOfPersons,
  toOthers: toCoTravellers,
  allPersons: getAllPersons,
  firstname: getFirstName,
  lastname: getLastName,
  gender: getGender,
  dob: getDateOfBirth,
  age: getAge,
  nationality: getNationality,
  countryName: getCountryName,
  frequentFlyerNumber: getFrequentFlyerNumber,
  ssrCodes: getSSRCodes,
};

export default PersonUtil;

export {
  getNationality,
  getCountryName,
  getDateOfBirth,
  getAge,
  getPerson,
  getGender,
  getFirstName,
  getLastName,
  toCoTravellers,
  getTotalNumberOfPersons,
  getOtherPersons,
  getFrequentFlyerNumber,
  getSSRCodes,
};
