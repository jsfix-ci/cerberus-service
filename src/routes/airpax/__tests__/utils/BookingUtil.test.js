import renderer from 'react-test-renderer';
import { BookingUtil } from '../../utils';
import { LONDON_TIMEZONE, UNKNOWN_TEXT } from '../../../../constants';

import config from '../../../../config';

describe('BookingUtil', () => {
  let booking;

  beforeEach(() => {
    config.dayjsConfig.timezone = LONDON_TIMEZONE;

    booking = {
      reference: 'LSV4UV',
      type: null,
      paymentMethod: null,
      bookedAt: '2020-02-07T17:15:00Z',
      checkInAt: null,
      ticket: {
        number: null,
        type: null,
        price: null,
      },
      country: 'GB',
      payments: [
        {
          amount: 2190.48,
          card: {
            number: '30XXXXXXXXXXX63X',
            expiry: '2020-10-01T00:00:00Z',
          },
        },
        {
          amount: 2151.48,
          card: {
            number: '30XXXXXXXXXXX62X',
            expiry: '2020-10-01T00:00:00Z',
          },
        },
      ],
      agent: {
        iata: '987654321',
        location: 'Galiway, IE',
      },
    };
  });

  it('should return a booking object if present', () => {
    const targetTaskMin = {
      movement: {
        booking: {
          reference: null,
          bookedAt: null,
          checkInAt: null,
          country: null,
        },
      },
    };

    const expected = {
      reference: null,
      bookedAt: null,
      checkInAt: null,
      country: null,
    };

    const output = BookingUtil.get(targetTaskMin);

    expect(output).not.toBeUndefined();
    expect(output).toEqual(expected);
  });

  it('should not return a booking object if not present', () => {
    const targetTaskMin = {
      movement: {
      },
    };

    const output = BookingUtil.get(targetTaskMin);

    expect(output).toBeNull();
  });

  it('should return the default error text when bookedAt is not present', () => {
    booking.bookedAt = null;
    const output = BookingUtil.toBookingText(booking);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return a formatted date if bookedAt is present', () => {
    booking.bookedAt = '1966-05-13T00:00:00Z';
    const output = BookingUtil.toBookingText(booking);
    expect(output).toEqual('13 May 1966');
  });

  it('should have bookedAt if populated', () => {
    booking.bookedAt = '1966-05-13T00:00:00Z';
    const output = BookingUtil.bookedAt(booking);
    expect(output).toEqual('1966-05-13T00:00:00Z');
  });

  it('should return null if bookedAt is assigned null', () => {
    booking.bookedAt = null;
    const output = BookingUtil.bookedAt(booking);
    expect(output).toBeNull();
  });

  it('should return check-in date if present', () => {
    booking.checkInAt = '1966-05-13T00:00:00Z';
    const output = BookingUtil.checkInAt(booking);
    expect(output).toEqual(booking.checkInAt);
  });

  it('should return unknown if check-in date is null', () => {
    booking.checkInAt = null;
    const output = BookingUtil.checkInAt(booking);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown if check-in date is undefined', () => {
    booking.checkInAt = undefined;
    const output = BookingUtil.checkInAt(booking);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown if check-in date is an empty string', () => {
    booking.checkInAt = '';
    const output = BookingUtil.checkInAt(booking);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return formatted check-in text if check-in time is present within the booking', () => {
    booking.checkInAt = '2020-08-07T17:15:00Z';
    const output = BookingUtil.toCheckInText(booking);
    expect(output).toEqual('Check-in 17:15');
  });

  it('should return default formatted error text if check-in time is not present within the booking', () => {
    booking.checkInAt = null;
    const output = BookingUtil.toCheckInText(booking);
    expect(output).toEqual(`Check-in ${UNKNOWN_TEXT}`);
  });

  it('should return a booking reference if present', () => {
    booking.reference = 'REF-12345';
    const output = BookingUtil.bookingRef(booking);
    expect(output).toEqual('REF-12345');
  });

  it('should return the default error text if booking reference is not present', () => {
    booking.reference = null;
    const output = BookingUtil.bookingRef(booking);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should calculate the booked prior value for given', () => {
    const bookedAt = '2020-08-07T17:15:00Z';
    const departureTime = '2020-08-10T18:15:00Z';

    const output = BookingUtil.bookedPrior(bookedAt, departureTime);
    expect(output).toEqual('3 days before travel');
  });

  it('should handle the omission of one the two dates', () => {
    const bookedAt = '';
    const departureTime = '2020-08-10T18:15:00Z';

    const output = BookingUtil.bookedPrior(bookedAt, departureTime);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should get the expected country code', () => {
    const output = BookingUtil.countryCode(booking);
    expect(output).toEqual('GB');
  });

  it('should return unknown country code is null', () => {
    booking.country = null;
    const output = BookingUtil.countryCode(booking);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when country code is invalid', () => {
    booking.country = 'UN';
    const output = BookingUtil.countryName(booking);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return the expected country name', () => {
    let output = BookingUtil.countryName(booking);
    expect(output).toEqual('United Kingdom');

    booking.country = 'FR';
    output = BookingUtil.countryName(booking);
    expect(output).toEqual('France');
  });

  it('should return unknown when the country name is null', () => {
    booking.country = null;
    let output = BookingUtil.countryName(booking);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return booking type when the booking type is present', () => {
    booking.type = 'Online';
    const type = BookingUtil.bookingType(booking);
    expect(type).toEqual(booking.type);
  });

  it('should return unknown when booking type is not present', () => {
    const type = BookingUtil.bookingType(booking);
    expect(type).toEqual(UNKNOWN_TEXT);
  });

  it('should return a booking ticket object if present', () => {
    const ticket = BookingUtil.bookingTicket(booking);

    const expected = {
      number: null,
      type: null,
      price: null,
    };

    expect(ticket).toEqual(expected);
  });

  it('should return null when a booking ticket is not present', () => {
    booking.ticket = null;
    const ticket = BookingUtil.bookingTicket(booking);
    expect(ticket).toBeNull();
  });

  it('should return the ticket number if present', () => {
    booking.ticket.number = 'TIC-098376';
    const ticketNumber = BookingUtil.ticketNumber(booking.ticket);
    expect(ticketNumber).toEqual(booking.ticket.number);
  });

  it('should return unknown when ticket number is not present', () => {
    const ticketNumber = BookingUtil.ticketNumber(booking.ticket);
    expect(ticketNumber).toEqual(UNKNOWN_TEXT);
  });

  it('should return the ticket type if present', () => {
    booking.ticket.type = 'Online';
    const ticketType = BookingUtil.ticketType(booking.ticket);
    expect(ticketType).toEqual(booking.ticket.type);
  });

  it('should return unknown when ticket type is not present', () => {
    const ticketType = BookingUtil.ticketType(booking.ticket);
    expect(ticketType).toEqual(UNKNOWN_TEXT);
  });

  it('should return the payment object if present', () => {
    const payments = BookingUtil.payments(booking);

    const expected = [
      {
        amount: 2190.48,
        card: {
          number: '30XXXXXXXXXXX63X',
          expiry: '2020-10-01T00:00:00Z',
        },
      },
      {
        amount: 2151.48,
        card: {
          number: '30XXXXXXXXXXX62X',
          expiry: '2020-10-01T00:00:00Z',
        },
      },
    ];

    expect(payments).toEqual(expected);
  });

  it('should return unknown when the payment object is not present', () => {
    booking.payments = null;
    const payments = BookingUtil.payments(booking);
    expect(payments).toBeNull();
  });

  it('should get a payment amount if present', () => {
    const paymentAmount = BookingUtil.paymentAmount(booking.payments[0]);
    expect(paymentAmount).toEqual(2190.48);
  });

  it('should return unknown when payment amount is not present', () => {
    booking.payments[0].amount = null;
    const paymentAmount = BookingUtil.paymentAmount(booking.payments[0]);
    expect(paymentAmount).toEqual(UNKNOWN_TEXT);
  });

  it('should return the payment card object', () => {
    const paymentCard = BookingUtil.paymentCard(booking.payments[0]);

    const expected = {
      number: '30XXXXXXXXXXX63X',
      expiry: '2020-10-01T00:00:00Z',
    };

    expect(paymentCard).toEqual(expected);
  });

  it('should return null if the payment card object is not present', () => {
    booking.payments[0] = null;

    const paymentCard = BookingUtil.paymentCard(booking.payments[0]);

    expect(paymentCard).toBeNull();
  });

  it('should return the payment card number', () => {
    const cardNumber = BookingUtil.paymentCard(booking.payments[0]).number;
    expect(cardNumber).toEqual(booking.payments[0].card.number);
  });

  it('should return the payment card expiry', () => {
    const cardExpiry = BookingUtil.paymentCard(booking.payments[0]).expiry;
    expect(cardExpiry).toEqual(booking.payments[0].card.expiry);
  });

  it('should return the payment card last 4 digits', () => {
    const cardLastFourDigits = BookingUtil.cardLastFourDigits(booking.payments[0]);
    const expected = booking.payments[0].card.number.substring(booking.payments[0].card.number.length - 4);
    expect(cardLastFourDigits).toEqual(expected);
  });

  it('should return unknown when the payment card\'s last 4 digits is not found', () => {
    booking.payments[0].card.number = null;
    const cardLastFourDigits = BookingUtil.cardLastFourDigits(booking.payments[0]);
    expect(cardLastFourDigits).toEqual(UNKNOWN_TEXT);
  });

  it('should return the payment card expiry date if present', () => {
    const cardExpiry = BookingUtil.cardExpiry(booking.payments[0]);
    expect(cardExpiry).toEqual('10/20');
  });

  it('should return unknown when the payment card expiry date is not present', () => {
    booking.payments[0].card.expiry = null;
    const cardExpiry = BookingUtil.cardExpiry(booking.payments[0]);
    expect(cardExpiry).toEqual(UNKNOWN_TEXT);
  });

  it('should get the agent object if present', () => {
    const agent = BookingUtil.agent(booking);

    const expected = {
      iata: '987654321',
      location: 'Galiway, IE',
    };

    expect(agent).toEqual(expected);
  });

  it('should get the agent iata if present', () => {
    const agentIata = BookingUtil.agentIata(booking.agent);
    expect(agentIata).toEqual(booking.agent.iata);
  });

  it('should return unknown when the agent iata is not present', () => {
    booking.agent.iata = null;
    const agentIata = BookingUtil.agentIata(booking.agent);
    expect(agentIata).toEqual(UNKNOWN_TEXT);
  });

  it('should get the agent location if present', () => {
    const agentLocation = BookingUtil.agentLocation(booking.agent);
    expect(agentLocation).toEqual(booking.agent.location);
  });

  it('should return unknown when the agent location is not present', () => {
    booking.agent.location = null;
    const agentLocation = BookingUtil.agentLocation(booking.agent);
    expect(agentLocation).toEqual(UNKNOWN_TEXT);
  });

  it('should render the payments block', () => {
    const section = renderer.create(BookingUtil.paymentsBlock(booking)).toJSON();
    expect(section).toMatchSnapshot();
  });
});
