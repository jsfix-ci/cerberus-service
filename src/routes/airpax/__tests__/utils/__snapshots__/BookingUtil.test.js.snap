// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingUtil should render the payments block 1`] = `
Array [
  <div
    className="bottom-border-thin govuk-!-margin-top-1 govuk-!-padding-bottom-1"
  >
    <div
      className="font__bold"
    >
      2190.48
    </div>
    <div
      className="font__light"
    >
      Credit card ending 
      X63X
      , expiry 
      10/20
    </div>
  </div>,
  <div
    className="bottom-border-thin govuk-!-margin-top-1 govuk-!-padding-bottom-1"
  >
    <div
      className="font__bold"
    >
      2151.48
    </div>
    <div
      className="font__light"
    >
      Credit card ending 
      X62X
      , expiry 
      10/20
    </div>
  </div>,
]
`;
